packages/std/src/math/decimal256.rs

```diff
@@ -0,0 +1,984 @@
+/// (which is (2^256 - 1) / 10^36)
+#[derive(Copy, Clone, Default, Debug, PartialEq, Eq, PartialOrd, Ord, JsonSchema)]
+pub struct Decimal256(#[schemars(with = "String")] Uint256);
+
+impl Decimal256 {
+    fn decimal_fractional() -> Uint256 {
+        // 1*10**36
+        Uint256::from(1_000_000_000_000_000_000_000_000_000_000_000_000u128)
+    }
+
+    fn decimal_fractional_squared() -> Uint256 {
+        // 1*10**72
+        Uint256::new([
+            0, 0, 144, 228, 15, 190, 234, 29, 58, 74, 188, 137, 85, 233, 70, 254, 49, 205, 207,
+            102, 246, 52, 225, 0, 0, 0, 0, 0, 0, 0, 0, 0,
+        ])
+    }
+
+    pub const MAX: Self = Self(Uint256::MAX);
+
+    /// Create a 1.0 Decimal256
+    pub fn one() -> Self {
+        Self(Self::decimal_fractional())
+    }
+
+    /// Create a 0.0 Decimal256
+    pub const fn zero() -> Self {
+        Self(Uint256::zero())
+    }
+
+    /// Convert x% into Decimal256
+    pub fn percent(x: u64) -> Self {
+        Self(Uint256::from(x) * Uint256::from(10_000_000_000_000_000_000_000_000_000_000_000u128))
+    }
+
+    /// Convert permille (x/1000) into Decimal256
+    pub fn permille(x: u64) -> Self {
+        Self(Uint256::from(x) * Uint256::from(1_000_000_000_000_000_000_000_000_000_000_000u128))
+    }
+
+    /// Returns the ratio (numerator / denominator) as a Decimal256
+    pub fn from_ratio(numerator: impl Into<Uint256>, denominator: impl Into<Uint256>) -> Self {
+        let numerator: Uint256 = numerator.into();
+        let denominator: Uint256 = denominator.into();
+        if denominator.is_zero() {
+            panic!("Denominator must not be zero");
+        }
+
+        Self(
+            // numerator * DECIMAL_FRACTIONAL / denominator
+            numerator.multiply_ratio(Self::decimal_fractional(), denominator),
+        )
+    }
+
+    pub fn is_zero(&self) -> bool {
+        self.0.is_zero()
+    }
+
+    /// Returns the approximate square root as a Decimal256.
+    ///
+    /// This should not overflow or panic.
+    pub fn sqrt(&self) -> Self {
+        // Algorithm described in https://hackmd.io/@webmaster128/SJThlukj_
+        // We start with the highest precision possible and lower it until
+        // there's no overflow.
+        (0..=18)
+            .rev()
+            .find_map(|i| self.sqrt_with_precision(i))
+            // The last step (i = 0) is guaranteed to succeed because `isqrt(Uint256::MAX) * 10^18` does not overflow
+            .unwrap()
+    }
+
+    /// Lower precision means more aggressive rounding, but less risk of overflow.
+    /// Precision *must* be a number between 0 and 36 (inclusive).
+    ///
+    /// Returns `None` if the internal multiplication overflows.
+    fn sqrt_with_precision(&self, precision: u32) -> Option<Self> {
+        let inner_mul = Uint256::from(100u128).pow(precision);
+        self.0.checked_mul(inner_mul).ok().map(|inner| {
+            let outer_mul = Uint256::from(10u128).pow(18 - precision);
```

Comment:
This probably needs to be adapted to the number of decimal places. Let's write a test that catches this bug.